array_converter mit Attributen konfigurieren

MHajoha opened the pull request:
- `$config->rename` wird ersetzt durch `#[array_key(...)]`
- `$config->alias` wird ersetzt durch `#[array_alias(...)]`
- `$config->array_elements` wird ersetzt durch `#[array_element_class(...)]`
- `$config->discriminated_by`, `$config->variant` und `$config->fallback_variant` wird ersetzt durch `#[array_polymorphic(...)]`

~~Es wird natürlich einige Konflikte mit <https://github.com/questionpy-org/moodle-qtype_questionpy/pull/133/commits/4e273bc6d6199edf1632d759ec2fcb929406f4e9> geben, die würde ich dann fixen, sobald das gemergt ist.~~

MHajoha committed:
refactor: use PHP 8 attributes to configure array_converter

MHajoha committed:
refactor: remove array_converter::configure and related methods

janbritz commented on classes/form/elements/checkbox_element.php:
```diff
@@ -70,7 +69,7 @@ public function __construct(string $name, ?string $leftlabel = null, ?string $ri
      * @param render_context $context target context
-     * @param int|null $group         passed by {@see checkbox_group_element::render_to} to the checkboxes belonging to
+     * @param int|null $group passed by {@see checkbox_group_element::render_to} to the checkboxes belonging to
      *                                it
```
```suggestion
     * @param int|null $group passed by {@see checkbox_group_element::render_to} to the checkboxes belonging to it
```

janbritz commented on classes/package/package.php:
```diff
@@ -56,9 +53,9 @@ class package extends package_base {
-                                ?array $tags = null) {
+        ?string $author = null, ?string $url = null, ?array $languages = null,
+        ?array $description = null, ?string $icon = null, ?string $license = null,
+        ?array $tags = null) {
```
Ich finde die Formatierung hier und bei den anderen `__construct`s deutlich unübersichtlicher, als die Vorherige. (Ist vermutlich auch automatisch passiert.)

MHajoha replied:
Ihh! Ja, war automatisch

janbritz reviewed (approved):
Ziemlich nice :D

larsbonczek reviewed (commented):
Du verwendest überall double quotes für Strings. Der Coding Style von Moodle erlaubt das eigentlich nur in Sonderfällen: https://moodledev.io/general/development/policies/codingstyle#strings

MHajoha committed:
style: add Squiz.Strings.DoubleQuoteUsage to phpcs.xml

Moodle prefers single-quoted strings.

See: https://moodledev.io/general/development/policies/codingstyle#strings

MHajoha committed:
style: have phpcbf replace string double-quotes with single-quotes

MHajoha commented:
> Du verwendest überall double quotes für Strings. Der Coding Style von Moodle erlaubt das eigentlich nur in Sonderfällen: https://moodledev.io/general/development/policies/codingstyle#strings

Cringe. Es passt eigentlich nicht in den PR, aber ich habe es in zwei neuen Commits für das ganze Repo angepasst (danke phpcbf). Kann ich gerne stattdessen auf dev machen, wenn euch das den PR zu sehr zumüllt.

MHajoha committed:
ci: work around moodlecheck bug

A dirty hack which applies https://github.com/moodlehq/moodle-local_moodlecheck/pull/148 during CI.
